server/controllers/finance/reports/debtors/debtorAccountBalance.js

```diff
@@ -0,0 +1,90 @@
+const _ = require('lodash');
+const ReportManager = require('../../../../lib/ReportManager');
+const db = require('../../../../lib/db');
+
+// path to the template to render
+const TEMPLATE = './server/controllers/finance/reports/debtors/balance.handlebars';
+
+exports.debtorAccountBalance = debtorAccountBalance;
+
+/**
+ * @method debtorAccountBalance
+ *
+ * @description
+ * The HTTP interface which actually creates the report.
+ */
+function debtorAccountBalance(req, res, next) {
+  const options = _.extend(req.query, {
+    filename : 'REPORT.CLIENT_DEBTOR_ACCOUNT_BALANCE_REPORT',
+    orientation : 'portrait',
+    csvKey   : 'rows',
+    footerRight : '[page] / [toPage]',
+    footerFontSize : '7',
+  });
+
+  let report;
+  let data;
+
+  try {
+    report = new ReportManager(TEMPLATE, req.session, options);
+  } catch (e) {
+    return next(e);
+  }
+
+  // fire the SQL for the report
+
+  const fiscalYearId = req.params.fiscalYearId;
+  let label = '';
+  queryContext(fiscalYearId)
+  .then(
+    function(data) {
+
+      fiscalYearQuery(fiscalYearId).then(function (response){
+
+        if(response.length>0){
+          label = response[0].label;
+        }
+        var _data = {rows : data, fiscalLabel : label };
+        report.render(_data)
+        .then(result => {
+          res.set(result.headers).send(result.report);
+        })
+      })
+      .catch(next)
+      .done();
+  })
+  .catch(next)
+  .done();
+}
+
+/**
+ * @method queryContext
+ *
+ * @param {Object} params Parameters passed in to customise the report - these
+ *                        are usually passed in through the query string
+ * @description
+ * The HTTP interface which actually creates the report.
+ */
+function queryContext(fiscalYearId) {
+
+  // aggregates the data above as totals into columns of 30, 60, 90, and >90
+  const sql = `
+    SELECT ac.id as account, ac.label as name, p.number as period,
+      SUM(pt.debit) as debit, SUM(pt.credit) as credit,
+      SUM(IF(p.number = 0,1,0))  as openingBalance,
+      SUM(IF(p.number <> 0, (pt.debit-pt.credit) ,0))  as closingBalance
```

Comment:
Did you test this?

It looks like the closing balance should be `SUM(pt.debit) - SUM(pt.credit)` instead of `pt.debit - pt.credit`.

Reply:
It was tested ,  if I try this : SUM(IF(p.number <> 0, ( SUM(pt.debit) - SUM(pt.credit) ) ,0)) ,an error occures

Reply:
Note that we are summing in a certain condition